### PR TITLE
Require CFFI on CPython for pure-Python operation.

### LINKED ISSUE #77: Can we require CFFI for PURE_PYTHON usage on CPython?
CFFI distributes binary wheels, so it's a lightweight dependency.

This would simplify our code by letting us remove one of the two implementations of picklecache.Ring, and it would simplify our test matrix and help us reach 100% coverage.

### PR DESCRIPTION
**Pull request by jamadden**
Fixes #77

### CODE REVIEW

**Comment by jamadden** [comment]
rebased on master to resolve conflicts with #106

File: setup.py [hunk 1]
```diff
@@ -131,6 +130,10 @@ def _read_file(filename):
       },
       install_requires=[
           'zope.interface',
+          "cffi ; platform_python_implementation == 'CPython'",
```
[the remark this answers is not in the document]
Comment (jamadden) [reply setup.py]: That requires a "recent" (2 years) setuptools. It's the (preferred, IIRC) replacement for the "empty extra" syntax.

**Comment by jamadden** [comment]
Thanks!

**Comment by dataflake** [comment]
@jamadden I know this is 7 years after the fact, but...

In #77 you were confident that CFFI is just a lightweight dependency and not problematic. From several years of experience we know this is not the case. It has been a never ending bugbear, especially around the time we are trying to implement support for newer Pythons and CFFI just doesn't keep up and breaks in unexpected ways. This is happening once again for 3.15 and we have to go through the many repositories that depend on `persistent` and remove testing for 3.15 until CFFI has caught up and we can enable it again. When that happens, who knows.

My suggestion would be to revert this PR at least partially to remove the dependency. Are there any really important reasons for this change other than the code simplification and test coverage you mentioned in #77?

I am volunteering to do this reversion because I am fed up with the constant issues.

**Comment by jamadden** [comment]
I completely agree, CFFI has become much heavier weight than expected, and I have no objections to seeing it made optional or removed entirely, so long as a PURE_PYTHON mode is still functional --- its major advantage was performance in that mode, and, aside from debugging, the main need for performance in PURE_PYTHON mode was PyPy, but my understanding is that PyPy isn't expecting to make releases for Python 3.12+ so its relevance is diminishing.

**Pull request merged**